File: app/src/main/java/al/edu/feut/financaime/BudgetFragment.java
package al.edu.feut.financaime;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.GridLayoutManager;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import al.edu.feut.financaime.adapter.LogAdapter;
import al.edu.feut.financaime.callback.DialogCallBack;
import al.edu.feut.financaime.dialog.EditBudget;
import al.edu.feut.financaime.dialog.EditIncomes;
import al.edu.feut.financaime.model.Budget;
import al.edu.feut.financaime.model.Database;
import al.edu.feut.financaime.util.Utilities;

public class BudgetFragment extends Fragment implements View.OnClickListener{

    private Budget mBudget;
    private AppCompatTextView budgetTv;
    private AppCompatTextView incomesTv;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mBudget = new Database(getContext()).budget(Utilities.month(Utilities.month()));
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_budget, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        AppCompatTextView editIncomes = view.findViewById(R.id.edit_incomes);
        AppCompatTextView editBudget = view.findViewById(R.id.edit_budget);

        budgetTv = view.findViewById(R.id.budget);
        incomesTv = view.findViewById(R.id.incomes);

        budgetTv.setText(Utilities.format(mBudget != null ? mBudget.getBudget() : 0));
        incomesTv.setText(Utilities.format(mBudget != null ? mBudget.getIncomes() : 0));

        if(mBudget == null) {
            editBudget.setOnClickListener(this);
            editBudget.setText(R.string.edit);
        } else if(mBudget.getIncomes() - mBudget.getExpense() == 0){
            editBudget.setOnClickListener(null);
            editBudget.setText(R.string.out_of_incomes_alert);
        } else {
            editBudget.setOnClickListener(this);
            editBudget.setText(R.string.edit);
        }
        editIncomes.setOnClickListener(mBudget == null ? this : null);

        RecyclerView log = view.findViewById(R.id.log);
        log.setLayoutManager(new GridLayoutManager(getActivity(), 4));
        log.setItemAnimator(new DefaultItemAnimator());
        log.setAdapter(new LogAdapter(new Database(getContext()).expense(Utilities.month(Utilities.month())), R.layout.log_single_item));
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.edit_incomes:
                new EditIncomes
                        .Builder()
                        .setBudget(mBudget)
                        .build()
                        .onDialogCallBack(new DialogCallBack() {
                            @Override
                            public void onClickInsert(Budget budget) {
                                if(budget.setId(new Database(getActivity()).insertBudget(budget) )!= -1) {
                                    budgetTv.setText(Utilities.format(budget.getBudget()));
                                    incomesTv.setText(Utilities.format(budget.getIncomes()));
                                    mBudget = budget;
                                }
                            }

                            @Override
                            public void onClickUpdate(Budget budget) {
                                if(new Database(getActivity()).updateIncomesValue(budget) != -1) {
                                    budgetTv.setText(Utilities.format(budget.getBudget()));
                                    incomesTv.setText(Utilities.format(budget.getIncomes()));
                                    mBudget = budget;
                                }
                            }
                        })
                        .show(getFragmentManager(), "INCOMES");
                break;
            case R.id.edit_budget:
                new EditBudget
                        .Builder()
                        .setBudget(mBudget)
                        .build()
                        .onDialogCallBack(new DialogCallBack() {
                            @Override
                            public void onClickInsert(Budget budget) {
                                if(budget.setId(new Database(getActivity()).insertBudget(budget)) != -1) {
                                    budgetTv.setText(Utilities.format(budget.getBudget()));
                                    incomesTv.setText(Utilities.format(budget.getIncomes()));
                                    mBudget = budget;
                                }
                            }

                            @Override
                            public void onClickUpdate(Budget budget) {
                                if(new Database(getActivity()).updateBudgetValue(budget) != 0) {
                                    budgetTv.setText(Utilities.format(budget.getBudget()));
                                    incomesTv.setText(Utilities.format(budget.getIncomes()));
                                    mBudget = budget;
                                }
                            }
                        })
                        .show(getFragmentManager(), "BUDGET");
                break;
        }
    }
}

File: app/src/main/java/al/bruno/personal/expense/BudgetFragment.kt
package al.bruno.personal.expense

import al.bruno.personal.expense.adapter.CustomAdapter
import al.bruno.personal.expense.callback.BindingData
import al.bruno.personal.expense.callback.OnClickListener
import al.bruno.personal.expense.callback.OnEditListeners
import android.os.Bundle
import androidx.fragment.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import al.bruno.personal.expense.databinding.FragmentBudgetBinding
import al.bruno.personal.expense.databinding.LogSingleItemBinding
import al.bruno.personal.expense.dialog.EditBudgetDialog
import al.bruno.personal.expense.dialog.EditIncomesDialog
import al.bruno.personal.expense.model.Budget
import al.bruno.personal.expense.model.BudgetMaster
import al.bruno.personal.expense.model.Expense
import al.bruno.personal.expense.util.Utilities.month
import al.bruno.personal.expense.view.model.BudgetMasterViewModel
import al.bruno.personal.expense.view.model.BudgetViewModel
import al.bruno.personal.expense.view.model.ExpenseViewModel
import android.util.Log
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers

class BudgetFragment : Fragment() {
    private val disposable : CompositeDisposable = CompositeDisposable()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val fragmentBudgetBinding: FragmentBudgetBinding = DataBindingUtil.inflate(inflater, R.layout.fragment_budget, container, false)
        ViewModelProviders.of(this)
                .get(BudgetMasterViewModel::class.java)
                .budget(month(month()))
                .observe(this, Observer{
                    run {
                        fragmentBudgetBinding.budget = it ?: BudgetMaster()
                    }
                }
        )
        fragmentBudgetBinding.onClickListenerEditBudget = object: OnClickListener<BudgetMaster> {
            override fun onClick(t: BudgetMaster) {
                EditBudgetDialog
                        .Builder()
                        .setBudget(budget = t)
                        .build()
                        .onDialogEditListeners(object : OnEditListeners<Budget> {
                            override fun onEdit(t: Budget) {
                                if(t.id == 0.toLong()) {
                                    disposable.add(ViewModelProviders
                                            .of(this@BudgetFragment)
                                            .get(BudgetViewModel::class.java)
                                            .insert(t)
                                            .subscribeOn(Schedulers.io())
                                            .subscribe({
                                                Log.i(BudgetFragment::class.java.name, it.toString())
                                            }, {
                                                Log.i(BudgetFragment::class.java.name, it.message)
                                            }))
                                } else {
                                    Thread(Runnable {
                                        ViewModelProviders
                                                .of(this@BudgetFragment)[BudgetViewModel::class.java]
                                                .updateBudget(t.budget, t.id)
                                    }).start()
                                }
                            }
                            override fun onDismiss(t: Budget) {
                            }
                        })
                        .show(fragmentManager, BudgetFragment::class.java.name)
            }
        }
        fragmentBudgetBinding.onClickListenerEditIncomes = object: OnClickListener<BudgetMaster> {
            override fun onClick(t: BudgetMaster) {
                EditIncomesDialog
                        .Builder()
                        .setBudget(budget = t)
                        .build()
                        .onDialogEditListeners(object : OnEditListeners<Budget>{
                            override fun onEdit(t: Budget) {
                                if(t.id == 0.toLong()) {
                                    disposable.add(ViewModelProviders
                                            .of(this@BudgetFragment)
                                            .get(BudgetViewModel::class.java)
                                            .insert(t)
                                            .subscribeOn(Schedulers.io())
                                            .subscribe({
                                                Log.i(BudgetFragment::class.java.name, it.toString())
                                            }, {
                                                Log.i(BudgetFragment::class.java.name, it.message)
                                            }))
                                } else {
                                    Thread(Runnable {
                                        ViewModelProviders
                                                .of(this@BudgetFragment)[BudgetViewModel::class.java]
                                                .updateIncomes(t.incomes, t.id)
                                    }).start()
                                }
                            }
                            override fun onDismiss(t: Budget) {
                            }

                        }).show(fragmentManager, BudgetFragment::class.java.name)
            }
        }
        ViewModelProviders.of(this)[ExpenseViewModel::class.java].expenses(month(month())).observe(this, Observer {
            run {
                fragmentBudgetBinding.logAdapter = CustomAdapter(it, R.layout.log_single_item, object : BindingData<Expense, LogSingleItemBinding> {
                    override fun bindData(t: Expense, vm: LogSingleItemBinding) {
                        vm.setExpense(t)
                    }
                })

            }
        })
        return fragmentBudgetBinding.root
    }

    override fun onStop() {
        super.onStop()
        disposable.clear()
    }
}
